File: src/com/kingmang/lazurite/utils/ValueUtils.java
package com.kingmang.lazurite.utils;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

import com.kingmang.lazurite.core.Function;
import com.kingmang.lazurite.core.Types;
import com.kingmang.lazurite.exceptions.LZRException;
import com.kingmang.lazurite.runtime.*;
import com.kingmang.lazurite.runtime.Types.*;
import lombok.NoArgsConstructor;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

@NoArgsConstructor
public final class ValueUtils {

    public static Object toObject(LzrValue val) throws JSONException {
        switch (val.type()) {
            case Types.ARRAY:
                return toObject((LzrArray) val);
            case Types.MAP:
                return toObject((LzrMap) val);
            case Types.NUMBER:
                return val.raw();
            case Types.STRING:
                return val.asString();
            default:
                return JSONObject.NULL;
        }
    }

    public static JSONObject toObject(LzrMap map) throws JSONException {
        final JSONObject result = new JSONObject(new LinkedHashMap<String, Object>());
        for (Map.Entry<LzrValue, LzrValue> entry : map) {
            final String key = entry.getKey().asString();
            final Object value = toObject(entry.getValue());
            result.put(key, value);
        }
        return result;
    }

    public static JSONArray toObject(LzrArray array) throws JSONException {
        final JSONArray result = new JSONArray();
        for (LzrValue value : array) {
            result.put(toObject(value));
        }
        return result;
    }

    public static LzrValue toValue(Object obj) throws JSONException {
        if (obj instanceof JSONObject) {
            return toValue((JSONObject) obj);
        }
        if (obj instanceof JSONArray) {
            return toValue((JSONArray) obj);
        }
        if (obj instanceof String) {
            return new LzrString((String) obj);
        }
        if (obj instanceof Number) {
            return LzrNumber.of(((Number) obj));
        }
        if (obj instanceof Boolean) {
            return LzrNumber.fromBoolean((Boolean) obj);
        }
        // NULL or other
        return LzrNumber.ZERO;
    }

    public static LzrMap toValue(JSONObject json) throws JSONException {
        final LzrMap result = new LzrMap(new LinkedHashMap<>(json.length()));
        final Iterator<String> it = json.keys();
        while(it.hasNext()) {
            final String key = it.next();
            final LzrValue value = toValue(json.get(key));
            result.set(new LzrString(key), value);
        }
        return result;
    }

    public static LzrArray toValue(JSONArray json) throws JSONException {
        final int length = json.length();
        final LzrArray result = new LzrArray(length);
        for (int i = 0; i < length; i++) {
            final LzrValue value = toValue(json.get(i));
            result.set(i, value);
        }
        return result;
    }

    public static Number getNumber(LzrValue value) {
        if (value.type() == Types.NUMBER) return ((LzrNumber) value).raw();
        return value.asInt();
    }

    public static float getFloatNumber(LzrValue value) {
        if (value.type() == Types.NUMBER) return ((LzrNumber) value).raw().floatValue();
        return (float) value.asNumber();
    }

    public static byte[] toByteArray(LzrArray array) {
        final int size = array.size();
        final byte[] result = new byte[size];
        for (int i = 0; i < size; i++) {
            result[i] = (byte) array.get(i).asInt();
        }
        return result;
    }

    public static Function consumeFunction(LzrValue value, int argumentNumber) {
        return consumeFunction(value, " at argument " + (argumentNumber + 1));
    }

    public static Function consumeFunction(LzrValue value, String errorMessage) {
        final int type = value.type();
        if (type != Types.FUNCTION) {
            throw new LZRException("TypeExeption ", "Function expected" + errorMessage
                    + ", but found " + Types.typeToString(type));
        }
        return ((LzrFunction) value).getValue();
    }

    public static <T extends Number> LzrMap collectNumberConstants(Class<?> clazz, Class<T> type) {
        LzrMap result = new LzrMap(20);
        for (Field field : clazz.getDeclaredFields()) {
            if (!Modifier.isStatic(field.getModifiers())) continue;
            if (!field.getType().equals(type)) continue;
            try {
                result.set(field.getName(), LzrNumber.of((T) field.get(type)));
            } catch (IllegalAccessException ignore) {
            }
        }
        return result;
    }
}

File: src/main/java/com/kingmang/lazurite/utils/ValueUtils.kt
package com.kingmang.lazurite.utils

import com.kingmang.lazurite.core.*
import com.kingmang.lazurite.core.Function
import com.kingmang.lazurite.runtime.values.*
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.lang.reflect.Modifier

object ValueUtils {
    @JvmStatic
    @Throws(JSONException::class)
    fun toObject(value: LzrValue): Any =
        when (value.type()) {
            Types.ARRAY -> value.asLzrArray().toJsonArray()
            Types.MAP -> value.asLzrMap().toJsonObject()
            Types.NUMBER -> value.asLzrNumber().raw()
            Types.STRING -> value.asLzrString().raw()
            else -> JSONObject.NULL
        }

    @JvmStatic
    @Throws(JSONException::class)
    fun toValue(obj: Any): LzrValue =
        when (obj) {
            is JSONObject -> obj.toLzrMap()
            is JSONArray -> obj.toLzrArray()
            is String -> LzrString(obj)
            is Number -> LzrNumber.of(obj)
            is Boolean -> LzrNumber.fromBoolean(obj)
            else -> LzrNumber.ZERO
        }

    @JvmStatic
    fun getNumber(value: LzrValue): Number =
        value.asLzrNumberOrNull()?.raw() ?: value.asInt()

    @JvmStatic
    fun getFloatNumber(value: LzrValue): Float =
        value.asLzrNumberOrNull()?.raw()?.toFloat() ?: value.asNumber().toFloat()

    @JvmStatic
    fun toByteArray(array: LzrArray): ByteArray =
        ByteArray(array.size()) { array[it].asInt().toByte() }

    @JvmStatic
    fun consumeFunction(value: LzrValue, argumentNumber: Int): Function =
        value.asLzrFunction { "Function expected at argument ${argumentNumber + 1}, but found ${Types.typeToString(value.type())}" }.value

    @Suppress("UNCHECKED_CAST")
    fun <T : Number> collectNumberConstants(clazz: Class<*>, type: Class<T>): LzrMap {
        val result = LzrMap(20)
        for (field in clazz.declaredFields) {
            if (!Modifier.isStatic(field.modifiers)) continue
            if (field.type != type) continue
            try {
                result[field.name] = LzrNumber.of(field[type] as T)
            } catch (ignore: IllegalAccessException) {
            }
        }
        return result
    }

    private fun LzrMap.toJsonObject(): JSONObject {
        val result = JSONObject(LinkedHashMap<String, Any>())
        forEach { (key, value) -> result.put(key.asString(), toObject(value)) }
        return result
    }

    private fun LzrArray.toJsonArray(): JSONArray {
        val result = JSONArray()
        asSequence().map(::toObject).forEach(result::put)
        return result
    }

    private fun JSONObject.toLzrMap(): LzrMap {
        val result = LzrMap(length())
        keys().forEach { result[LzrString(it)] = toValue(get(it)) }
        return result
    }

    private fun JSONArray.toLzrArray(): LzrArray =
        LzrArray(length()) { toValue(get(it)) }
}